File: LenientBenchmark/TreeAccumulator.cs
using System.Collections.Generic;
using System.Threading.Tasks;

namespace LenientBenchmark
{
    public static class TreeAccumulator
    {
        public static List<T> AccumulateLeaves<T>(Tree<T> tree)
        {
            var l = new List<T>();
            AccumulateLeaves(tree, l);
            return l;
        }

        private static void AccumulateLeaves<T>(Tree<T> tree, List<T> leaves)
        {
            if (tree is Leaf<T> leaf)
            {
                leaves.Add(leaf.Value);
            }
            else
            {
                var node = tree as Node<T>;

                AccumulateLeaves(node.Left, leaves);
                AccumulateLeaves(node.Right, leaves);
            }
        }

        public static List<T> AccumulateLeavesForkJoin<T>(Tree<T> tree)
        {
            var l = new List<T>();
            var t = AccumulateLeavesForkJoin(tree, l);
            t.Wait();
            return l;
        }

        private static async Task AccumulateLeavesForkJoin<T>(Tree<T> tree, List<T> leaves)
        {
            if (tree is Leaf<T> leaf)
            {
                leaves.Add(leaf.Value);
            }
            else
            {
                var node = tree as Node<T>;

                var left = Task.Run(async () =>
                {
                    var leftLeaves = new List<T>();
                    await AccumulateLeavesForkJoin(node.Left, leftLeaves);
                    return leftLeaves;
                });
                var right = Task.Run(async () =>
                {
                    var rightLeaves = new List<T>();
                    await AccumulateLeavesForkJoin(node.Right, rightLeaves);
                    return rightLeaves;
                });
                var l = await left;
                leaves.AddRange(l);
                var r = await right;
                leaves.AddRange(r);
            }
        }
        
        public static List<T> AccumulateLeavesLenient<T>(Tree<T> tree)
        {
            var l = Task.Run(() => new List<T>());
            var t = AccumulateLeavesLenient(Task.FromResult(tree), l);
            Task.WaitAll(l, t);
            return l.Result;
        }

        private static async Task AccumulateLeavesLenient<T>(Task<Tree<T>> tree, Task<List<T>> leaves)
        {
            var t = await tree;
            if (t is Leaf<T> leaf)
            {
                var l = await leaves;
                l.Add(leaf.Value);
            }
            else
            {
                var n = t as Node<T>;

                var left = Task.Run(async () =>
                {
                    await AccumulateLeavesLenient(Task.FromResult(n.Left), leaves);
                    return leaves.Result;
                });
                var right = Task.Run(async () =>
                {
                    await AccumulateLeavesLenient(Task.FromResult(n.Right), left);
                });

                await Task.WhenAll(left, right);
            }
        }
    }
}

File: Fsharp/Problems/TreeAccumulator.fs
module Fsharp.Problems.TreeAccumulator

open Fsharp.TreeGenerator
open System.Threading.Tasks

let leaves a =
    let rec leavesAccum n leaves =
        match n with
        | (Leaf i) -> i :: leaves
        | Node (left, right, _) -> leavesAccum left (leavesAccum right leaves)
    leavesAccum a []

let lazyLeaves a =
    let lst = lazy(leaves a)
    lst.Value

let parallelLeaves t =
    let rec asyncAccum (t:tree) = async {        
        match t with
        | (Leaf i) -> return [i]
        | Node (left, right, _) -> return [(asyncAccum left); (asyncAccum right)]
                                       |> Async.Parallel
                                       |> Async.RunSynchronously
                                       |> List.concat
    }
    Async.RunSynchronously (asyncAccum t)

let tplParallelLeaves t = 
    let rec leavesAccum n =
        match n with
        | (Leaf i) -> [i] 
        | Node (left, right, _) ->
            let l = Task.Factory.StartNew<int64 list> (fun () ->  leavesAccum left)
            let r = Task.Factory.StartNew<int64 list> (fun () ->  leavesAccum right)
            Task.WaitAll(l, r)
            List.append l.Result r.Result
                                
    leavesAccum t

let lazyParallel t =
    let lst = lazy(parallelLeaves t)
    lst.Value
    
let lazyTPLLeaves t =
    let lst = lazy(tplParallelLeaves t)
    lst.Value
    
let lenientLeaves t =
    let rec lenientLeavesAccum n = async {
        match n with
        | Leaf i -> return [i]
        | Node (left, right, _) ->
            let! l = Async.StartChild (lenientLeavesAccum left)
            let! r = Async.StartChild (lenientLeavesAccum right)
            let! lr = l
            let! rr = r
            return List.concat [lr ; rr]
    }
    
    Async.RunSynchronously (lenientLeavesAccum t)